lib/Serialization/Serialization.cpp

```diff
@@ -585,6 +590,25 @@ serialization::TypeID Serializer::addTypeRef(Type ty) {
   return TypesToSerialize.addRef(ty);
 }
 
+serialization::ClangTypeID Serializer::addClangTypeRef(const clang::Type *ty) {
+  if (ty) {
+    // Try to serialize the non-canonical type, but fall back to the
+    // canonical type if necessary.
+    auto loader = getASTContext().getClangModuleLoader();
+    if (!ty->isCanonicalUnqualified()) {
+      if (!loader->isSerializable(ty, /*check canonical*/ false)) {
+        ty = ty->getCanonicalTypeInternal().getTypePtr();
+      }
+    }
+    if (!loader->isSerializable(ty, /*check canonical*/ false)) {
```

Comment:
The logic here seems a bit hard to follow. For example, this looks like if we have a non-canonical type, then we will do a second check for `isSerializable` even if the first check passed, which is probably not intended. Instead, you could refactor it to something like:

```
bool isSerializable;
if (ty->isCanonicalUnqualified()) {
  isSerializable = loader->isSerializable(ty, /*checkCanonical*/ false);
} else if (loader->isSerializable(ty, /*checkCanonical*/ false)) {
  isSerializable = true;
} else {
  ty = ty->getCanonicalTypeInternal().getTypePtr();
  isSerializable = loader->isSerializable(ty, /*checkCanonical*/ false);
}
if (!isSerializable) {
  // report fatal error
}
```

This might seem a bit too pedestrian (e.g. I could've gotten rid of the `isSerializable = true;` by initializing it with `true` at the beginning instead and having a `!loader->isSerializable` to compress it further) but the benefit IMO is that there is only 1 negated condition at the end, all the others are positive checks so it is easy to see that the value is initialized in all the branches.

Reply:
Reworked